core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -646,10 +646,9 @@ private[spark] class ExecutorAllocationManager(
   private[spark] class ExecutorAllocationListener extends SparkListener {
 
     private val stageIdToNumTasks = new mutable.HashMap[Int, Int]
-    // Number of running tasks per stage including speculative tasks.
-    // Should be 0 when no stages are active.
-    private val stageIdToNumRunningTask = new mutable.HashMap[Int, Int]
     private val stageIdToTaskIndices = new mutable.HashMap[Int, mutable.HashSet[Int]]
+    private val liveTaskIds = new mutable.HashSet[Long]
```

Comment:
Which problem does this change aim to address? Do we have different expectation for 'regular' task and speculative task?

Reply:
To simplify the calculation of totalRunningTasks (```add remove size```), instead of using the previous commit, use ```taskIdToNumLiveTasks = new mutable.HashMap[Long, Int] (add i++ i-- remove), values.sum```.
And in the case of many tasks can improve the sum calculation speed.
The final purpose is to prevent the totalRunningTasks from becoming negative, resulting in the job stuck, and can further improve the efficiency of applying for resources.

Reply:
The taskid is actually not repeated.
Only the same taskid appears in some test cases, and the task has different types.
```scala
createTaskInfo(0, 0, "executor-1")
createTaskInfo(0, 0, "executor-2", true)
```

https://github.com/apache/spark/blob/cd4a284030e5142bdb405ff5b71735ac8cee2dde/core/src/test/scala/org/apache/spark/ExecutorAllocationManagerSuite.scala#L260-L267